File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;

        // iterate over all grid spaces
        for (int row = 0; row < grid.length; row++) {
            for (int col = 0; col < grid[0].length; col++) {

                // get area of island, dfs traversal
                int area = getIslandArea(grid, row, col);

                // set max area to this island if larger than current max size
                if (area > maxArea) {
                    maxArea = area;
                }
            }
        }
        return maxArea;
    }

    private int getIslandArea(int[][] grid, int row, int col) {
        // return if out of bounds
        if (row < 0 || col < 0 || row >= grid.length || col >= grid[0].length) {
            return 0;
        }

        // return if not land
        if (grid[row][col] == 0) {
            return 0;
        }

        // set land to explored
        // do not revisit
        grid[row][col] = 0;

        // return sum of this land and all connected land
        return 1 + getIslandArea(grid, row + 1, col) + getIslandArea(grid, row - 1, col) + getIslandArea(grid, row, col + 1) + getIslandArea(grid, row, col - 1);
    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.kt
class Solution {
    fun maxArea(heights: IntArray): Int {
        var max = 0

        var p1 = 0
        var p2 = heights.size - 1

        while (p1 < p2) {
            val height = minOf(heights[p1], heights[p2])
            val width = p2 - p1
            val sum = width * height

            max = maxOf(sum, max)

            if (heights[p1] < heights[p2]) {
                p1++
            } else {
                p2--
            }
        }

        return max
    }
}
